File: 1-Part/4-chapter/preorderTraversal.py
root = {
    'data': 'A', 'children': [
        {'data': 'B', 'children': [
            {'data': 'D', 'children': []}
        ]},
        {'data': 'C', 'children': [
            {'data': 'E', 'children': [
                {'data': 'G', 'children': []},
                {'data': 'H', 'children': []}
            ]},
            {'data': 'F', 'children': []}
        ]}
    ]
}


def preorderTraverse(node):
    print(node['data'], end=' ')
    if len(node['children']) > 0:
        for child in node['children']:
            preorderTraverse(child)
    return


preorderTraverse(root)

File: 1-Part/4-chapter/preorderTraversal.js
root = {
    'data': 'A', 'children': [
        {
            'data': 'B', 'children': [
                {'data': 'D', 'children': []}
            ]
        },
        {
            'data': 'C', 'children': [
                {
                    'data': 'E', 'children': [
                        {'data': 'G', 'children': []},
                        {'data': 'H', 'children': []}
                    ]
                },
                {'data': 'F', 'children': []}
            ]
        }
    ]
}

function preorderTraverse(node) {
    console.log(node['data'] + '')

    if (node['children'].length > 0) {
        for (let i = 0; i < node['children'].length; i++) {
            preorderTraverse(node['children'][i])
        }
    }
    return
}

preorderTraverse(root)